File: quickstart_guides/arrays_strings/python/is_anagram.py
"""
Title: Is Anagram

Problem: 
    Given two strings s and t , write a function to determine if t is an
    anagram of s.

Execution: python is_anagram.py
"""
import unittest


def is_anagram_sorting(s: str, t: str) -> bool:
    """Check if string s is anagram of string t using sorting."""
    if len(s) != len(t):
        return False
    s = sorted(s)
    t = sorted(t)
    return s == t


def is_anagram_hash_table(s: str, t: str) -> bool:
    """Check if string s is anagram of string t using hash table."""
    ht = dict()

    if len(s) != len(t):
        return False

    for i in s:
        if i in ht:
            ht[i] += 1
        else:
            ht[i] = 1
    for i in t:
        if i in ht:
            ht[i] -= 1
        else:
            ht[i] = 1
    for i in ht:
        if ht[i] != 0:
            return False
    return True


class TestIsAnagram(unittest.TestCase):
    """Unit test for is_anagram."""

    def test_1(self):
        s = "anagram"
        t = "nagaram"
        self.assertEqual(is_anagram_sorting(s, t), True)
        self.assertEqual(is_anagram_hash_table(s, t), True)

    def test_2(self):
        s = "rat"
        t = "car"
        self.assertEqual(is_anagram_sorting(s, t), False)
        self.assertEqual(is_anagram_hash_table(s, t), False)


if __name__ == "__main__":
    unittest.main()

File: quickstart_guides/arrays_strings/java/IsAnagram.java
/*
 *   Title: Is Anagram
 *
 *   Problem:
 *      Given two strings s and t , write a function to determine if t is an
 *      anagram of s.
 *
 *   Execution: javac IsAnagram.java && java IsAnagram
 */

import java.util.Arrays;
import java.util.List;

public class IsAnagram {
    public static boolean isAnagramSort(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        char[] str1 = s.toCharArray();
        char[] str2 = t.toCharArray();
        Arrays.sort(str1);
        Arrays.sort(str2);
        return Arrays.equals(str1, str2);
    }

    public static boolean isAnagramHashTable(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        int[] counter = new int[26];
        for (int i = 0; i < s.length(); i++) {
            counter[s.charAt(i) - 'a']++;
            counter[t.charAt(i) - 'a']--;
        }
        for (int count : counter) {
            if (count != 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        assert isAnagramSort("anagram", "nagaram") == true;
        assert isAnagramSort("rat", "car") == false;

        assert isAnagramHashTable("anagram", "nagaram") == true;
        assert isAnagramHashTable("rat", "car") == false;
        System.out.println("Passed all test cases");
    }
}
